ssds/checksum.py

```diff
@@ -28,3 +28,44 @@ def compute_composite_etag(etags: typing.List[str]) -> str:
     bin_md5 = b"".join([binascii.unhexlify(etag) for etag in etags])
     composite_etag = md5(bin_md5).hexdigest() + "-" + str(len(etags))
     return composite_etag
+
+class UnorderedChecksum:
+    def __init__(self):
+        raise NotImplementedError()
+
+    def update(self, chunk_number: int, data: bytes):
+        raise NotImplementedError()
+
+    def hexdigest(self) -> str:
+        raise NotImplementedError()
+
+class S3EtagUnordered(UnorderedChecksum):
+    def __init__(self):
+        self._checksums: List[Tuple[int, str]] = list()
+
+    def update(self, chunk_number: int, data: bytes):
+        self._checksums.append((chunk_number, md5(data).hexdigest()))
+
+    def hexdigest(self) -> str:
+        return compute_composite_etag([cs[1] for cs in sorted(self._checksums)])
+
+class GScrc32cUnordered(UnorderedChecksum):
+    def __init__(self):
+        self._current_chunk_number = 0
+        self._chunks: Dict[int, bytes] = dict()
+        self._checksum = crc32c()
+
+    def update(self, chunk_number: int, data: bytes):
```

Comment:
Instead of:

```
    def update(self, chunk_number: int, data: bytes):
        self._chunks[chunk_number] = data
        for chunk_number in sorted(self._chunks.keys()):
            if self._current_chunk_number == chunk_number:
                self._checksum.update(self._chunks[chunk_number])
                del self._chunks[chunk_number]
                self._current_chunk_number += 1
            else:
                break
```

I think that this is equivalent, and it makes it clearer for me at least:

```
    def update(self, chunk_number: int, data: bytes):
        self._chunks[chunk_number] = data
        if self._current_chunk_number in self._chunks:
            self._checksum.update(self._chunks[self._current_chunk_number])
            del self._chunks[self._current_chunk_number]
            self._current_chunk_number += 1
```

Also addressed in the suggestion above, but "chunk_number" is an arg input, and it's overwritten in the iterator, which is a nit but it's a potential gotcha missing that it was overwritten.

Reply:
The loop is needed to eagerly consume as many chunks as possible, which help lower memory footprint.

Will fix the overwritten parameter.

Reply:
Ack, I see it.  I guess that that's not equivalent.  Marking as resolved.